torch/_inductor/cudagraph_trees.py

```diff
@@ -324,40 +341,57 @@ class StorageWeakRefWrapper:
     Wrapper around a storage weak ref. Will deallocate it upon expiration if invoked.
     """
 
-    __slots__ = ["ref", "_data_ptr"]
+    __slots__ = ["ref", "_data_ptr", "extra_ref_check"]
 
     storage_ref: Optional[StorageWeakRef]
 
-    def __init__(self, inp: Union[Tensor, UntypedStorage]):
+    def __init__(
+        self,
+        inp: Union[Tensor, UntypedStorage],
+        extra_ref_check: Optional[Callable[[], None]] = None,
+    ):
         if isinstance(inp, Tensor):
             stor = inp.untyped_storage()
         else:
             assert isinstance(inp, UntypedStorage)
             stor = inp
         self.ref = StorageWeakRef(stor)
         self._data_ptr = stor.data_ptr()
+        self.extra_ref_check = extra_ref_check
 
     @classmethod
-    def from_weakref_and_data_ptr(cls, cdata, data_ptr):
+    def from_weakref_and_data_ptr(cls, cdata, data_ptr, extra_ref_check=None):
         instance = cls.__new__(cls)
         instance._data_ptr = data_ptr
         instance.ref = StorageWeakRef.from_weakref(cdata)
+        instance.extra_ref_check = extra_ref_check
         return instance
 
     def __call__(self) -> Optional[StorageWeakRefPointer]:
-        if self.ref is None:
-            return None
-
-        if self.ref.expired():
-            self.ref = None
+        if self.expired():
             return None
 
         return self.ref.cdata
 
+    def swap_weakref(self, cdata):
+        self.ref.__del__()
+        self.ref.cdata = cdata
```

Comment:
Why do you overwrite the StorageWeakRef directly? This seems extremely un-kosher. You wouldn't overwrite a Python weakref.

Reply:
StorageWeakRef is a wrapper around a StorageImpl. It's different than a python weakref. This is to avoid doing extra allocations on the hot path (it did actually show up on profiles).